source/common/network/connection_impl.cc

```diff
@@ -106,10 +112,28 @@ void ConnectionImpl::close(ConnectionCloseType type) {
 
     closeSocket(ConnectionEvent::LocalClose);
   } else {
-    // TODO(mattklein123): We need a flush timer here. We might never get open socket window.
-    ASSERT(type == ConnectionCloseType::FlushWrite);
-    close_with_flush_ = true;
+    ASSERT(type == ConnectionCloseType::FlushWrite ||
+           type == ConnectionCloseType::FlushWriteAndDelay);
+    delayed_close_ = true;
+    bool delayed_close_timeout_set = delayedCloseTimeout().count() > 0;
+
+    // All close types that follow do not actually close() the socket immediately so that buffered
+    // data can be written. However, we do want to stop reading to apply TCP backpressure.
     read_enabled_ = false;
+
+    // Force a closeSocket() after the write buffer is flushed if the close_type calls for it or if
+    // no delayed close timeout is set.
+    close_after_flush_ = !delayed_close_timeout_set || type == ConnectionCloseType::FlushWrite;
+
+    // Create and activate a timer which will immediately close the connection if triggered.
+    // A config value of 0 disables the timeout.
+    if (delayed_close_timeout_set) {
+      delayed_close_timer_ = dispatcher_.createTimer([this]() -> void { onDelayedCloseTimeout(); });
+      ENVOY_CONN_LOG(debug, "setting delayed close timer with timeout {} ms", *this,
+                     delayedCloseTimeout().count());
+      delayed_close_timer_->enableTimer(delayedCloseTimeout());
+    }
```

Comment:
1. Yes, I am addressing an outstanding `TODO(mattklein123)` to create a timer to terminate the connection if the flush never completes.  It is possible that `closeSocket(LocalClose)` will be called twice if the timer event has fired when the write flush completes, but the second call returns immediately due to the `if (fd() == -1)` check.

2. I am adding a runtime check to avoid processing a delayed close request twice, since it does seem to be possible for `close()` to be called multiple times with FlushWrite/FlushWriteAndDelay based on tests I just ran.

Reply:
Done.